File: Numerical_Methods/Solving Ordinary Differential Equations (ODEs)/Runge-Kutta Methods (RK)/python_solutions/problem4.py
import numpy as np

def f(x, y):
    return np.exp(x) - y  # y' = e^x - y

def rk3(x0, y0, h, target_x):
    x = x0
    y = y0
    
    while x < target_x:
        k1 = h * f(x, y)
        k2 = h * f(x + h/2, y + k1/2)
        k3 = h * f(x + h, y - k1 + 2*k2)
        
        y = y + (k1 + 4*k2 + k3) / 6
        x = x + h
    
    return y

def main():
    x0 = 0
    y0 = 1
    h = 0.2
    target_x = 0.6
    
    result = rk3(x0, y0, h, target_x)
    print(f"Solution at x = {target_x:.1f}: y = {result:.6f}")
    
    # Exact solution: y = (x*e^x + y0)
    exact = target_x*np.exp(target_x) + y0
    print(f"Exact solution: y = {exact:.6f}")
    print(f"Error: {abs(result - exact):.6f}")

if __name__ == "__main__":
    main() 

File: Numerical_Methods/Solving Ordinary Differential Equations (ODEs)/Runge-Kutta Methods (RK)/java_solutions/Problem4.java
public class Problem4 {
    public static double f(double x, double y) {
        return Math.exp(x) - y;  // y' = e^x - y
    }

    public static double rk3(double x0, double y0, double h, double targetX) {
        double x = x0;
        double y = y0;
        
        while (x < targetX) {
            double k1 = h * f(x, y);
            double k2 = h * f(x + h/2, y + k1/2);
            double k3 = h * f(x + h, y - k1 + 2*k2);
            
            y = y + (k1 + 4*k2 + k3) / 6;
            x = x + h;
        }
        
        return y;
    }

    public static void main(String[] args) {
        double x0 = 0;
        double y0 = 1;
        double h = 0.2;
        double targetX = 0.6;
        
        double result = rk3(x0, y0, h, targetX);
        System.out.printf("Solution at x = %.1f: y = %.6f%n", targetX, result);
        
        // Exact solution: y = (x*e^x + y0)
        double exact = targetX*Math.exp(targetX) + y0;
        System.out.printf("Exact solution: y = %.6f%n", exact);
        System.out.printf("Error: %.6f%n", Math.abs(result - exact));
    }
} 